techcity/events/connectors/youtube.py

```diff
@@ -0,0 +1,150 @@
+            if group_slug in group_slugs
+        }
+
+    def fetch(self, cached: bool) -> None:
+        print("Fetching event recordings from YouTube...")
+
+        for group_slug, playlist_id in self.event_recording_sources.items():
+            if cached:
+                print(f"Using cached data for {group_slug}...")
+            else:
+                self.fetch_to_cache(group_slug, playlist_id)
+
+            self.generate_event_recordings(group_slug, playlist_id)
+
+    def fetch_to_cache(self, group_slug: str, playlist_id: str):
+        """
+        Fetch the data from YouTube and save it to the cache.
+
+        Uses the following:
+            https://developers.google.com/youtube/v3/docs/playlistItems/list
+        """
+        retries = Retry(
+            total=3,
+            allowed_methods={"GET"},
+            status_forcelist=[502, 503, 504],
+            backoff_factor=0.1,
+        )
+        session = requests.Session()
+        session.mount("https://", HTTPAdapter(max_retries=retries))
+
+        results = []
+        paginating = True
+        page_token = ""
+
+        while paginating:
+            response = session.get(
+                "https://www.googleapis.com/youtube/v3/playlistItems",
+                params={
+                    "part": ["id", "contentDetails", "snippet", "status"],
+                    "playlistId": playlist_id,
+                    "maxResults": 50,
+                    "pageToken": page_token,
+                    "key": self.youtube_api_key,
+                },
+                headers={
+                    "Accept": "application/json",
+                    "Referer": "https://techfrederick.org",
+                },
+                timeout=5,
+            )
+            response.raise_for_status()
+            data = response.json()
+            results.extend(data["items"])
+            if "nextPageToken" in data:
+                page_token = data["nextPageToken"]
+            else:
+                paginating = False
+
+        with open(self.cache_dir / f"{group_slug}-youtube-videos.json", "w") as f:
+            json.dump(results, f)
+
+    def generate_event_recordings(self, group_slug: str, playlist_id: str) -> None:
+        """Generate any event recordings found in API data."""
+        group = Group.objects.get(slug=group_slug)
+        with open(self.cache_dir / f"{group_slug}-youtube-videos.json") as f:
+            youtube_video_data = json.load(f)
+
+        for video_data in youtube_video_data:
+            self.upsert_event_recording(group, video_data, playlist_id)
+
+    def upsert_event_recording(
+        self,
+        group: Group,
+        youtube_video_data: dict,
+        playlist_id: str,
+    ) -> None:
+        """Upsert an event recording into the database."""
+
+        snippet = youtube_video_data["snippet"]
+        content_details = youtube_video_data["contentDetails"]
+
+        # TODO - could extract Speaker ? Should probably be from the
```

Comment:
I don't reliably set the speaker in a consistent way, so I agree that we should probably get speaker data from a different source.